feat(profile): login with Keycloak

ardasnturk opened the pull request:
This PR allows the user to log in with Keycloak and save the resulting accessToken to the device. In order to test it, a new build needs to be created due to the native requirements of expo-web-browser.

To test:

To activate the profile property, firstly, the `withProfile` value in the `ScreenName.Home` object in the `defaultStackConfig` must be updated to `true`

Then the keys required for the login process must be added to the `globalSettings.settings` object as follows.

```
  "profile": {
    "clientId": "",
    "clientSecret": "",
    "scopes": [],
    "serverUrl": ""
  },
```

|not logged user|login alert|login with web browser|logged in user|
|--|--|--|--|
![Simulator Screenshot - iPhone 16 Pro Max - 2025-04-16 at 13 08 01](https://github.com/user-attachments/assets/a4b430ec-3764-4860-a016-7e775f168ba3)|![Simulator Screenshot - iPhone 16 Pro Max - 2025-04-16 at 13 08 04](https://github.com/user-attachments/assets/d64b086d-89d3-4d61-ba87-e85404491b53)|![Simulator Screenshot - iPhone 16 Pro Max - 2025-04-16 at 13 09 22](https://github.com/user-attachments/assets/5ea50f7b-bd43-4c78-81fb-4cc01e0148f7)|![Simulator Screenshot - iPhone 16 Pro Max - 2025-04-16 at 13 08 10](https://github.com/user-attachments/assets/b67df2c0-3f2e-47f7-a784-542a656016e5)

SVA-1511

ardasnturk committed:
feat: login with keykloak

- added `useLoginProfile` hook to control user's login actions
- added `LoginHeader` to show in the header on `HomeScreen` if `hasProfile` is activated on main-server
- added `expo-auth-session` and `expo-web-browser` packages for login processes
  - `expo-auth-session` (https://docs.expo.dev/versions/latest/sdk/auth-session) requires `expo-web-browser` (https://docs.expo.dev/versions/latest/sdk/webbrowser)
- added `refreshToken` function to `useLoginProfile` hook to refresh token with `refreshToken` after `accessToken` expires
- added `logout` function to `useLoginProfile` hook to allow user to logout

SVA-1511

donni106 commented on src/config/consts.js:
```diff
@@ -43,6 +43,10 @@ export const consts = {
     infoProvider: 'Anbieterinformationen (Taste)',
     lastName: 'Nachname',
     link: '(Link)',
+    login: 'Anmelden (Taste)',
+    loginHint: 'Zu den Anmeldedaten wechseln',
+    logout: 'Abmelden (Taste)',
+    logoutHint: 'Zu den Anmeldedaten wechseln',
```
is it intended to habe the same texts for `loginHint` and `logoutHint`? and where does `Zu den Anmeldedaten wechseln` come from? i don't get the meaning correctly, what should it tell someone?

ardasnturk replied:
Yes, you are right. loginHint and logoutHint should not be the same. These texts were added here for accessibility. If the user is logged in, the device reads logoutHint to the user.

donni106 commented on src/components/LoginHeader.tsx:
```diff
@@ -0,0 +1,49 @@
+
+  const { isLoggedIn, loading, login, logout } = useLoginProfile();
+
+  if (!hasProfile || loading) {
+    return null;
+  }
```
checking `hasProfile` should not happen here but already considered in the stack config boolean flag `withLogin`

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,183 @@
+  revocationEndpoint: serverUrl + '/revoke'
+};
+
+const CLIENT_ID = profile?.clientId;
+const CLIENT_SECRET = profile?.clientSecret;
```
what about using these values from global settings to be able to change them on the run instead of having to make new releases as they are shipped in the secrets file?

ardasnturk replied:
Yeah, you're right. It can be done. We've done it before. But I have some fears about it. Like security. After all, I need clientId and secretKey here and if we get them from `globalSettings`, there may be a security problem, right? After all, this request comes to the app as a result of a request and these requests can be easily intercepted nowadays.

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,183 @@
+      clientId: CLIENT_ID,
+      codeChallengeMethod: CodeChallengeMethod.S256,
+      redirectUri,
+      responseType: ResponseType.Code,
```
why do we add it here, if it is the default? 

![image](https://github.com/user-attachments/assets/fd34d716-f218-4273-924f-fd27b2eb3cd8)

https://docs.expo.dev/versions/latest/sdk/auth-session/#authrequestconfig

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,183 @@
+      redirectUri,
+      responseType: ResponseType.Code,
+      scopes: SCOPES,
+      usePKCE: false
```
why do we set this to `false`?

donni106 replied:
where do you know that from?

ardasnturk replied:
Because the pkce value is off in keycloak. But I don't know why it's off.

<img width="312" alt="image" src="https://github.com/user-attachments/assets/af53c344-8d1d-442a-aa83-86fc967e29ee" />

You can see this setting by following the route below.
keycloak -> Identity providers -> ITEBO

donni106 replied:
https://oauth.net/2/pkce, `true` per default: https://docs.expo.dev/versions/latest/sdk/auth-session/#authrequestconfig

so if someone changes this in keycloak, we need a new app release with also an adjusted value. can we consider this to be part of the global settings?

ardasnturk replied:
https://github.com/smart-village-solutions/smart-village-app-app/pull/1311/commits/ddabdcba94f62db4dbb1ad567f24403cc2e39ef8

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,183 @@
+          clientSecret: CLIENT_SECRET,
+          redirectUri,
+          extraParams: {
+            grant_type: 'authorization_code'
```
why is this hardcoded if everything else is set per variables?

ardasnturk replied:
Because it is used 1 time in the code, I did not feel the need to create an extra variable.

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,183 @@
+    }
+  };
+
+  useEffect(() => {
+    (async () => {
+      await checkToken();
+    })();
+  }, []);
```
wouldn't `useEffect(checkToken, []);` work here? why would we need to await something if we are not doing anything with its returned value?

ardasnturk replied:
This only enables the checkToken function to work. checkToken checks whether there is a token stored in the device and whether this token is expired. If the token has expired, a new token is automatically generated and stored in the device. 

But yes, you are right, if we don't need the value returned from the function, then the async/await structure can be removed.

ardasnturk replied:
In my opinion, I see no reason why this code should not work. But if I update the code as you say, I get this error. Therefore, I think it should remain as it is now.

<img width="461" alt="image" src="https://github.com/user-attachments/assets/b8cb76f2-2fdb-493a-b9ba-ea55fe50c424" />

donni106 replied:
huh, those are the same things. weird. and nothing has to do with some `create`, right? can you share the git diff of that change?

ardasnturk replied:
I only have this modification. I have no idea why the error message mentions `create`.

<img width="545" alt="image" src="https://github.com/user-attachments/assets/0075513b-d84d-4404-baa0-722c1887d201" />


The issue was caused by passing an async function directly to useEffect like this: 
```
  useEffect(checkToken, []);
```
Since React doesn’t allow async functions as effect callbacks (because they return a Promise), it led to a runtime error.That's why we have to leave the code as it is.

ardasnturk commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,183 @@
+  const [request, response, promptAsync] = useAuthRequest(
+    {
+      clientId: CLIENT_ID,
+      codeChallengeMethod: CodeChallengeMethod.S256,
```
It can be removed. I overlooked it.

ardasnturk commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,183 @@
+
+      if (now >= expiresAt) {
+        // Token expired, try refresh
+        const refreshed = await refreshToken(parsedToken.refreshToken);
```
I'm replacing it with `refreshedToken`.

ardasnturk commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,183 @@
+  }, []);
+
+  const checkToken = async () => {
+    const storedToken = await readFromStore(PROFILE_ACCESS_TOKEN);
```
I'm replacing it with `storedTokenData`.

ardasnturk committed:
feat(profile): make the profile property updatable from main-server

- added `profile` as a prop to `useLoginProfile.ts` so that the profile property can be used via `globalSettings`
- added `profile` values from `globalSettings` obtained in `LoginHeader.tsx` to `useLoginProfile` as a prop
- renamed `withLogin` prop to `withProfile` for clarity
- updated accessibility texts

SVA-1511

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,177 @@
+const PROFILE_ACCESS_TOKEN = 'profileAccessToken';
+
+type TProfile = {
+  serverUrl: string;
+  clientId: string;
+  clientSecret: string;
+  scopes: string[];
```
do we want them sorted alphabetically?

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,177 @@
+  const [loading, setLoading] = useState(false);
+
+  const redirectUri = makeRedirectUri({
+    scheme: appJson.expo.scheme,
+    path: 'redirect'
```
do we want them sorted alphabetically?

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,177 @@
+      );
+
+      setIsLoggedIn(true);
+      await addToStore(PROFILE_ACCESS_TOKEN, JSON.stringify(tokenResult));
```
shouln't we use the `SecureStore` with `SecureStore.setItemAsync(...)` for the token?

ardasnturk replied:
Yeah, you're right. I'll update it with `SecureStore`. 👍

donni106 replied:
getting this now

> Value being stored in SecureStore is larger than 2048 bytes and it may not be stored successfully. In a future SDK version, this call may throw an error.

ardasnturk replied:
We are currently using the `issueAt`, `expiresIn` and `refreshToken` values from the `tokenResult` object we saved in `SecureStore`. Different keys can be created for these 3 values and saved differently. We can also adapt the `token` value, which I think we will need in the future.

donni106 replied:
Only `refreshToken` needs to be put in `SecureStore`, right? The others can be in `AsyncStorage`. What do you think is the effort to refactor this?

ardasnturk replied:
I developed this code a long time ago. Therefore, I must first carefully analyse the changes that need to be made. I estimate it will take about two hours.

donni106 committed:
feat(profile): update logout function to use idToken for logout request

- changed the logout function to check for idToken instead of accessToken
- modified the logout request to include id_token_hint in the URL
  - saw that here https://keycloak.discourse.group/t/how-is-logout-really-working-and-can-we-bypass-the-logout-confirm-page/15314
- added `dismissAuthSession` to kill the browser session in iOS https://docs.expo.dev/versions/latest/sdk/webbrowser/#webbrowserdismissbrowser

SVA-1511

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,178 @@
+        }
+      });
+
+      await addToStore(PROFILE_ACCESS_TOKEN, '');
```
if not using secure store, this should be `removeFromStore`, right?

ardasnturk replied:
Yes 🙌 But I'm already updating with `SecureStore` now, so I'll use `SecureStore.deleteItemAsync(...)`.

ardasnturk committed:
feat(profile): add alert for logout

- added alert to notify users when logging out of their account

SVA-1511

ardasnturk committed:
feat(profile): update async storage with secure store

- updated with `SecureStore` instead of async storage to store user's token information
- props alphabetised

SVA-1511

ardasnturk committed:
feat(profile): make the `usePKCE` property updateable from main-server

SVA-1511

ardasnturk committed:
feat(profile): update token handling in `useLoginProfile`

- replaced `PROFILE_TOKEN` with `PROFILE_META` for better clarity
- added `addToStore` and `removeFromStore` functions for managing access metadata
- updated token exchange and check logic to utilize the new refresh token structure
- ensured logout process clears the correct tokens and metadata
- moved hooks to the bottom to improve readability

SVA-1511

donni106 commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    try {
+      const parsedTokenMeta = JSON.parse(storedAccessMeta);
```
now "access meta" is called "token meta"? please keep the namings in line.

ardasnturk commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,204 @@
+
+      setIsLoggedIn(true);
+
+      const { refreshToken: refresh, issueAt, expiresIn } = tokenResult;
```
I chose this because of the name conflict. However, I will fix it.

ardasnturk commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,204 @@
+import { addToStore, readFromStore, removeFromStore } from '../helpers';
+
+const PROFILE_REFRESH_TOKEN = 'profileRefreshToken';
+const PROFILE_ACCESS_META = 'profileAccessMeta';
```
I named it this way because the metadata returned from the server during login does not contain a token. I will update the naming. I don't like it this way.

I will change it to `PROFILE_TOKEN` and `PROFILE_META`.

ardasnturk commented on src/hooks/useLoginProfile.ts:
```diff
@@ -0,0 +1,204 @@
+      const parsedTokenMeta = JSON.parse(storedAccessMeta);
+
+      const now = Math.floor(Date.now() / 1000);
+      const expiresAt = parsedTokenMeta.issuedAt + parsedTokenMeta.expiresIn;
```
You've made a very good point. I've made a naming error here. The name of the data stored in Asyncstorage should also be `issuedAt`.

donni106 committed:
refactor(profile): login files

- updated destructuring of `globalSettings` in `ConfigurationsProvider` to ensure `settings` is always an object
- removed unused constant `LOGIN_MODAL` from `InfoHeader` and `LoginHeader` that also duplicated the one used in `LoginModal`
- ensured `settings` is always an object for consistency across all files where used
- adjusted destructuring in `useLoginProfile` to use `newRefreshToken` to avoid variable shadowing
- modified dependencies in `useEffect` hooks for preventing issue with stale closures

SVA-1511

donni106 committed:
feat(profile): update token handling in `useLoginProfile`

- errors won't be caught by the surrounding try-catch, and the operation may not complete before the function returns
- added `await`s to ensure proper error handling

SVA-1511
